ion-schema/src/isl/ranges.rs

```diff
@@ -0,0 +1,502 @@
+        fn is_empty(start: &Limit<T>, end: &Limit<T>) -> bool {
+            match (start, end) {
+                (Limit::Inclusive(lower), Limit::Inclusive(upper)) => lower > upper,
+                (Limit::Exclusive(lower), Limit::Exclusive(upper))
+                | (Limit::Exclusive(lower), Limit::Inclusive(upper))
+                | (Limit::Inclusive(lower), Limit::Exclusive(upper)) => lower >= upper,
+                _ => false,
+            }
+        }
+    }
+
+    /// Checks if two limits would result in an empty range of integers. Returns true if there are
+    /// no integer values between `start` and `end`, taking into consideration the exclusivity of
+    /// the limits.
+    pub fn is_int_range_empty<T: CheckedAdd + One + PartialOrd>(
+        start: &Limit<T>,
+        end: &Limit<T>,
+    ) -> bool {
+        match (start, end) {
+            (Limit::Inclusive(lower), Limit::Inclusive(upper)) => lower > upper,
+            (Limit::Exclusive(lower), Limit::Inclusive(upper))
+            | (Limit::Inclusive(lower), Limit::Exclusive(upper)) => lower >= upper,
+            (Limit::Exclusive(lower), Limit::Exclusive(upper)) => {
+                // Checking for e.g. range::[exclusive::1, exclusive::2] which is empty.
+                let adjusted_lower = lower.checked_add(&T::one());
+                // If the _lower_ bound wraps around when we add one, then we know it's empty.
+                if adjusted_lower.is_none() {
+                    return true;
+                }
+                adjusted_lower.unwrap() >= *upper
+            }
+            _ => false,
+        }
+    }
+
+    /// Represents an interval of values where the upper and lower ends can be open, closed, or unbounded.
+    ///
+    /// At least one of the ends must be [`Limit::Exclusive`] or [`Limit::Inclusive`]. A `Range` may not be
+    /// empty (i.e. there must be at least one value for which [`contains`] returns `true`).
+    #[derive(Debug, Clone, PartialEq)]
+    pub struct Range<T> {
+        lower: Limit<T>,
+        upper: Limit<T>,
+    }
+
+    // Note the trait bound! This allows us to inject a different non-empty check depending on the
+    // realized type of `T`.
+    impl<T: PartialOrd + Clone> Range<T>
+    where
+        Self: RangeValidation<T>,
+    {
+        /// Creates a new range.
+        /// At least one limit must be bounded, and the range must be non-empty.
+        pub fn new(start: Limit<T>, end: Limit<T>) -> IonSchemaResult<Self> {
+            isl_require!(start != Limit::Unbounded || end != Limit::Unbounded  => "range may not contain both 'min' and 'max'")?;
+            isl_require!(!Self::is_empty(&start, &end) => "")?;
+            Ok(Self {
+                lower: start,
+                upper: end,
+            })
+        }
+
+        /// Creates a new range with inclusive endpoints.
+        /// [start] must be less than or equal to [end].
+        pub fn new_inclusive(start: T, end: T) -> IonSchemaResult<Self> {
```

Comment:
(Suggestion) I should have probably brought this up in the first revision but should we name this `new_*()` methods without the new keyword. e.g. (`new_inclusive()` -> `inclusive()`)? For constraints/types I have used similar convention (e.g. `IslConstraint::byte_length()`, `IslType::named()`). Thoughts?

Reply:
😅 I actually think that `IslConstraint::byte_length()` should cease to exist. Instead, to get a byte length variant of `IslConstraint`, you should use `IslConstraint::ByteLength(ByteLength::new(...))` or `ByteLength::new(...).into()`. Same thing with `IslType`.

That being said, if you feel strongly about getting rid of the `new_` prefix, I will do it. I just figured it would be a helpful way to signal that it is a constructor.

Reply:
I was trying to follow this [design pattern example](https://rust-unofficial.github.io/patterns/patterns/creational/builder.html)(although this is for a builder pattern ): 
```rust
impl FooBuilder {
    pub fn new(/* ... */) -> FooBuilder {
        // Set the minimally required fields of Foo.
        FooBuilder {
            ...
        }
    }

    pub fn name(mut self, bar: String) -> FooBuilder {
        // Set the name on the builder itself, and return the builder by value.
        ...
    }
}
```
I don't think this needs to block the PR though, but I do think whatever we decide to do it should be consistent overall. I have created [this issue](https://github.com/amazon-ion/ion-schema-rust/issues/197) to make changes for it according to what we decide.